.pipelines/ci.yml

```diff
@@ -25,12 +25,16 @@ variables:
     value: arosvcdev.azurecr.io/vpn
   - name: LOCAL_E2E_IMAGE
     value: arosvcdev.azurecr.io/e2e
-  - name: TAG
-    value: $(Build.BuildId)
   - name: VERSION
-    value: $(Build.BuildId)
+    value: $(TAG)
+  - name: PR_TAG
+    value: $[ format('pr-{0}-{1}', variables['System.PullRequest.PullRequestId'], variables['System.PullRequest.SourceCommitId']) ]
+  - name: MASTER_TAG
+    value: $[ format('master-{0}', variables['Build.SourceVersion']) ]
+  - name: TAG
+    value: $[ coalesce(variables['PR_TAG'], variables['MASTER_TAG']) ]
```

Comment:
I think this still won't work. Let's assume we're on master, meaning the `System.PullRequest` attributes are null. `PR_TAG` will be set to `"pr--"`, and then `TAG` will try:

```
coalesce("pr--", "master-whatever-Build.SourceVersion-is")
```

which will return `"pr--"`.

I think one option would be to conditionally assign a value to `PR_TAG` such that it only gets a non-empty value when we're on a pull request branch. That way `coalesce(variables['PR_TAG'], variables['MASTER_TAG'])` will work how we want it to. It looks like there are some things worth trying here: https://stackoverflow.com/questions/57532138/can-conditional-variable-assignment-be-done-in-azure-pipelines

Reply:
I have added the code that  TAG generation to avoid "pr--" by ensuring PR_TAG is assigned only for PR builds. Implements conditional logic to validate System.PullRequest attributes, ensuring coalesce correctly falls back to MASTER_TAG for master builds, addressing concerns about invalid tags.